037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,251 @@
+Having a full auto-rebalancing is out of scope for this proposal, the field is added for enabling a future implementation.
+
+The template to be used is referenced by a new mandatory `spec.kafka.autoRebalance.template` field and the cluster operator uses that template to create a corresponding `KafkaRebalance` custom resource.
+If the referenced `KafkaRebalanceTemplate` custom resource doesn't exist, the cluster operator will fail the reconcile.
+
+In the following example, a user can decide to have different rebalancing templates when adding or removing brokers and when they want a full rebalancing during the normal cluster execution.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+    autoRebalance:
+      # using the same rebalancing template when adding or removing brokers
+      - mode: add-brokers
+        template: my-add-remove-brokers-rebalancing-template
+      - mode: remove-brokers
+        template: my-add-remove-brokers-rebalancing-template  
+      # using a different template for a full rebalancing  
+      - mode: full
+        template: my-full-rebalancing-template
+```
+
+The user can even decide to not have full auto-rebalancing and, for example, using different templates when adding or removing brokers.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+    autoRebalance:
+      # using a specific template when adding brokers
+      - mode: add-brokers
+        template: my-add-brokers-rebalancing-template
+      # using a different template when removing brokers  
+      - mode: remove-brokers
+        template: my-remove-brokers-rebalancing-template
+      # NOTE: no full auto-rebalancing   
+```
+
+The user could even decide to have auto-rebalancing only when adding or removing brokers but not for both.
+
+This way provides the greatest flexibility to describe in which case to run the auto-rebalancing and with which configuration.
+
+### Auto-rebalancing on scaling up
+
+Following the operator flow when auto-rebalancing on scaling up.
+
+1. The user scales up the cluster by increasing the `spec.kafka.replicas` field.
+2. The `KafkaReconciler` takes care of it by rolling the existing brokers first and then scaling up the cluster as usual, by adding additional brokers.
+3. If the `spec.kafka.autoRebalance` contains an entry with `add-brokers` mode, the `KafkaReconciler` creates a `KafkaRebalance` custom resource named as `auto-<cluster-name>-rebalance` by using the corresponding template and sets:
+   * the `spec.mode` as `add-brokers` and the `spec.brokers` with the list of brokers already added to the cluster.
+   * the `spec.autoApproval` from the template as described by this [proposal](https://github.com/strimzi/proposals/pull/56).
```

Comment:
I left other comment about the`autoApproval` ... but what if it doesn't have the field?

Reply:
This sentence is stale now, because we decided to not have "manual" approval anymore, so the  `autoApproval` will be set on the `KafkaRebalance` automatically and it doesn't need to be in the template first. I will fix this.